### PR TITLE
Add definition to remaining types

### PR DESCRIPTION
**Pull request by ishaanshah**
# Description
Adds definition to remaining types.

# Problem
Some types were defined as `any`, which doesn't help in development or to catch errors.

# Solution
Adding correct definitions to these types fixes the problem

### CODE REVIEW

File: listenbrainz/webserver/static/js/src/APIService.test.ts [hunk 1]
```diff
@@ -70,29 +102,90 @@ describe("submitListens", () => {
...
+        },
+      ])
     ).resolves.toEqual({
       ok: true,
       status: 200,
     });
   });
 
   it("calls itself recursively if size of payload exceeds MAX_LISTEN_SIZE", async () => {
-    // Change MAX_LISTEN_SIZE to 0
```
Comment (ishaanshah) [inline listenbrainz/webserver/static/js/src/APIService.test.ts:87]: `MAX_LISTEN_SIZE` can't be 0 as it would lead to infinite recursion.

File: listenbrainz/webserver/static/js/src/RecentListens.tsx [hunk 2]
```diff
@@ -206,6 +212,7 @@ export default class RecentListens extends React.Component<
         return {
           playingNowByUser: {
             ...prevState.playingNowByUser,
+            // @ts-ignore
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/RecentListens.tsx:215]: what was the error here?

File: listenbrainz/webserver/static/js/src/RecentListens.tsx [hunk 3]
```diff
@@ -410,7 +417,7 @@ export default class RecentListens extends React.Component<
                       .map((listen) => {
                         return (
                           <tr
-                            key={listen}
+                            key={listen.listened_at}
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/RecentListens.tsx:420]: let's use `listened_at-recording_msid-username` here. `listened_at` can still be duplicated, for example in follow page.

File: listenbrainz/webserver/static/js/src/types.d.ts [hunk 4]
```diff
@@ -2,8 +2,43 @@ declare module "spotify-web-playback-sdk";
 declare module "react-bs-notifier";
 declare module "time-ago";
 
-// TODO: make this type specific
-declare type Listen = any;
+// TODO: Remove "| null" when backend stops sending fields with null
+declare type Listen = {
+  listened_at: number;
+  listened_at_iso?: string | null;
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/types.d.ts:8]: Is there a way to get rid of these `| null`s, they seem non-ideal.

File: listenbrainz/webserver/static/js/src/types.d.ts [hunk 5]
```diff
@@ -2,8 +2,43 @@ declare module "spotify-web-playback-sdk";
...
+      release_msid?: string | null;
+      spotify_album_artist_ids?: Array<string> | null;
+      spotify_album_id?: string | null;
+      spotify_artist_ids?: Array<string> | null;
+      spotify_id?: string | null;
+      tags?: Array<string> | null;
+      track_mbid?: string | null;
+      tracknumber?: number | null;
+      work_mbids?: Array<string> | null;
+      [key: string]: any;
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/types.d.ts:37]: if you make additional info a seperate interface, you might not need this line.

File: listenbrainz/webserver/static/js/src/types.d.ts [hunk 6]
```diff
@@ -13,38 +48,75 @@ declare type ListenBrainzUser = {
...
   access_token: string;
   permission: string;
 };
 
-declare type SpotifyPermission = any; // TODO
-declare type SpotifyTrack = any; // TODO
-declare type SpotifyArtist = any; // TODO
-declare type SpotifyDeviceID = any; // TODO
-declare interface SpotifyImage {
+declare type SpotifyPermission = string;
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/types.d.ts:62]: we only have a few specific permissions for spotify. 

Let's make the type reflect that. (https://github.com/metabrainz/listenbrainz-server/blob/5b1a66ea/listenbrainz/domain/spotify.py#L14)

File: listenbrainz/webserver/static/js/src/types.d.ts (hunk 4 above)
Comment (ishaanshah) [reply listenbrainz/webserver/static/js/src/types.d.ts:8]: We can turn off `strictNullChecks` in typescript which will treat `null` and `undefined` as the same everywhere. But then we will also lose other errors such as **Possibly null/undefined**.

File: listenbrainz/webserver/static/js/src/RecentListens.tsx (hunk 2 above)
Comment (ishaanshah) [reply listenbrainz/webserver/static/js/src/RecentListens.tsx:215]: Fixed it!

File: listenbrainz/webserver/static/js/src/types.d.ts (hunk 5 above)
Comment (ishaanshah) [reply listenbrainz/webserver/static/js/src/types.d.ts:37]: I did not understand how making `additional_info` into a separate interface will help solve the issue. Can you please explain how it can be done.

File: listenbrainz/webserver/static/js/src/types.d.ts (hunk 4 above)
Comment (paramsingh) [reply listenbrainz/webserver/static/js/src/types.d.ts:32]: That does not sound like a good idea. Let's keep the nulls for now.

File: listenbrainz/webserver/static/js/src/types.d.ts [hunk 7]
```diff
@@ -13,38 +49,81 @@ declare type ListenBrainzUser = {
...
+  name: string;
+};
+
+declare type SpotifyTrack = {
+  album: {
+    uri: string;
+    name: string;
+    images: Array<SpotifyImage>;
+  };
+  artists: Array<SpotifyTrack>;
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/types.d.ts:87]:
```suggestion
  artists: Array<SpotifyArtist>;
```

?

File: listenbrainz/webserver/static/js/src/SpotifyPlayer.tsx [hunk 8]
```diff
@@ -82,6 +82,7 @@ export default class SpotifyPlayer extends React.Component<
     this.state = {
       accessToken: props.spotifyUser.access_token,
       permission: props.spotifyUser.permission,
+      // @ts-ignore
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/SpotifyPlayer.tsx:85]: why is this needed?

File: listenbrainz/webserver/static/js/src/Scrobble.ts [hunk 9]
```diff
@@ -57,7 +57,7 @@ export default class Scrobble {
     return "";
   }
 
-  scrobbledAt(): string {
+  scrobbledAt(): number {
     /* Returns scrobbledAt if present, else returns an empty string */
```
Comment (paramsingh) [inline listenbrainz/webserver/static/js/src/Scrobble.ts:61]: let's update the comment here too, we return a -1 now.

File: frontend-test.sh [hunk 10]
```diff
@@ -19,3 +19,7 @@ fi
 docker-compose -f docker/docker-compose.test.yml \
                -p listenbrainz_test \
                run --rm frontend_tester npm test
+
+docker-compose -f docker/docker-compose.test.yml \
```
Comment (paramsingh) [inline frontend-test.sh:23]: why did you add this back? We removed it in #790. If we want to keep it, I think frontend.sh should have options that allow for running just the tests or just the type checker.

Comment (ishaanshah) [reply frontend-test.sh:27]: I added it so that, type checking can be done on all files once before commit rather than opening each file and checking for errors. I have added an option to run type checker only, using `./frontend-test.sh -t`.

**Comment by paramsingh** [comment]
For the future, can you make changes after a review in a different commit without rebasing / amending the previous commit? It makes things easier to review.
